content/client-lib-development-guide/features.textile

```diff
@@ -565,6 +571,18 @@ h3(#realtime-channel). Channel
 * @(RTL14)@ If an @ERROR ProtocolMessage@ is received for this channel (the channel attribute matches this channel's name), then the channel should immediately transition to the FAILED state, and the @Channel.errorReason@ should be set
 * @(RTL16)@ @Channel#setOptions@ takes a @ChannelOptions@ object and sets or updates the stored channel options.
 ** @(RTL16a)@ If the user has provided either @ChannelOptions.params@ or @ChannelOptions.modes@ and the channel is in either the @attached@ or @attaching@ state, @Channel#setOptions@ sends an @ATTACH@ message to the server with the params & modes encoded per "@RTL4@":#RTL4, and indicates success once the server has replied with an @ATTACHED@ (or indicates failure if the channel becomes detached or failed before that happens, as with @Channel#attach@); else it indicates success immediately
+* @(RTL17)@ From the time that the channel enters @ATTACHING@ state, no further received messages are passed to subscribers until after the channel next enters the @ATTACHED@ state. Any messages received after the event that triggered the transition to @ATTACHING@, and prior to the transition to @ATTACHED@, must be discarded.
```

Comment:
It seems odd to special-case the `attaching` state in the spec. I know that that was our motivation for making the change, but I don't think the general client library API should special-case that state over all the other states that the user doesn't expect to receive messages. I'd argue that a better form of this spec would be something like `* @(RTL17)@ Message listeners should only be called if the channel is in the @ATTACHED@ state. Any messages received after when the channel is in any state other than @ATTACHED@ should be discarded`

Reply:
TBH I do not think it makes much of a difference in terms of what has to be implemented. I am fine with your edit, too. However, I think the wording here comes from @paddybyers so maybe he has a comment on this?

Also, is it correct to talk about "messages received"? The messages might be received before entering the `ATTACHING` state and they still will have to be discarded.

Reply:
> Also, is it correct to talk about "messages received"? The messages might be received before entering the ATTACHING state and they still will have to be discarded.

Fair enough, happy to go with `No messages should be passed to subscribers if the channel is in any state other than @ATTACHED@`.

The main reason I prefer this is it feels weird for the spec to theoretically be fine with a message being emitted while `detached`, `failed`, `suspended` etc., but not while `attaching`. Even if we know that can never happen, it feels wrong that just `attaching` is singled-out for this. (One case where the change will make a difference is the `detaching` state: with your version you can still have messages emitted during `detaching`, which feels inconsistent; in mine you won't.)

Reply:
Ok fixed now.